src/overlay/TCPPeer.cpp

```diff
@@ -343,14 +346,65 @@ TCPPeer::startRead()
         return;
     }
 
+    const size_t HDRSZ = 4;
     auto self = static_pointer_cast<TCPPeer>(shared_from_this());
 
     assert(self->mIncomingHeader.size() == 0);
 
     if (Logging::logTrace("Overlay"))
         CLOG(TRACE, "Overlay") << "TCPPeer::startRead to " << self->toString();
 
-    self->mIncomingHeader.resize(4);
+    self->mIncomingHeader.resize(HDRSZ);
+
+    // We read large-ish (1MB) buffers of data from TCP which might have quite a
+    // few messages in them. We want to digest as many of these _synchronously_
+    // as we can before we issue an async_read against ASIO.
+    YieldTimer yt(mApp.getClock());
+    while (mSocket->in_avail() >= HDRSZ && yt.shouldKeepGoing())
+    {
+        size_t n = mSocket->read_some(asio::buffer(mIncomingHeader));
+        if (n != HDRSZ)
+        {
+            drop("error during header read_some",
+                 Peer::DropDirection::WE_DROPPED_REMOTE,
+                 Peer::DropMode::IGNORE_WRITE_QUEUE);
+            return;
+        }
+        size_t length = static_cast<size_t>(getIncomingMsgLength());
+        if (length != 0)
+        {
+            if (mSocket->in_avail() >= length)
+            {
+                // We can finish reading a full message here synchronously.
+                mIncomingBody.resize(length);
+                n = mSocket->read_some(asio::buffer(mIncomingBody));
+                if (n != length)
+                {
+                    drop("error during body read_some",
```

Comment:
I would call `receivedBytes(n, false)` here

Reply:
yeah, this is all consolidated now. will post update.